- adds support for localized electric emission factors

baywet opened the pull request:
This pull requests adds support for country and state specific emission factors so the user's data can be more accurate.
It relies on the following data to work https://www.carbonfootprint.com/docs/2019_06_emissions_factors_sources_for_2019_electricity.pdf
I've only ported Canadian emission factors to get early feedback on the pull request before I invest too much time. We'd need to import the rest of the dataset.

Although the code is in place to support state specific, the library I'm using doesn't seem to support returning provinces/states at the moment. I have opened an issues over there to ask them if they are planning to add the information. https://github.com/ideditor/country-coder/issues/19
I made a point to use an offline library an not any API to avoid dependency on an API key and lengthy network calls that could slow down the app when importing data.
If they don't add that support, we could either look into other libraries to bridge the gaps or lookup by country only (less precise).

TODO:
- [x] feedback on the current approach
- [x] import remaining factors

corradio commented:
Hi @baywet - thanks for the PR. This looks really good, and is definitely an improvement over what we've done. The approach looks good to me. Let me know when you're ready for the final review and I'll make a final pass 👍

baywet commented:
@corradio thanks for the feedback. I was actually wrapping up the PR.
- the main geoloc reverse lookup library I was using is not planning to add states
- I found another library that does states for USA, Canada and Australia which I implemented in combination with the previous one
- added all the emission factors that were listed in the document

This PR is ready to be merged whenever you have a chance :)

corradio commented on co2eq/electricity/emissionfactors.json:
```diff
@@ -0,0 +1,122 @@
+    { "country": "USA", "state": "WI", "factor": 663.2 },
+    { "country": "USA", "state": "WV", "factor": 945.3 },
+    { "country": "USA", "state": "WY", "factor": 966.7 },
+    { "country": "AUT", "state": "", "factor": 142 },
```
`"state": null` is probably better

baywet replied:
fixed

corradio commented on co2eq/electricity/emissionfactors.json:
```diff
@@ -0,0 +1,122 @@
+{
+  "factors": [
+    { "country": "CAN", "state": "QC", "factor": 1.5 },
```
do you have access to the year of the data source? Might be relevant to put it in the json

baywet replied:
added

baywet commented:
@corradio thanks for merging! when do you think I'll see it on my phone?

corradio commented:
We'll try to make a release within next week! Stay tuned (or ask us on Slack)

baywet commented:
@corradio I just got the update on my phone and I saw the bubbles resizing and the progress bar updating live! Really cool!

corradio commented:
Great!

On Thu, May 28, 2020 at 3:28 AM Vincent Biret <notifications@github.com>
wrote:

> @corradio <https://github.com/corradio> I just got the update on my phone
> and I saw the bubbles resizing and the progress bar updating live! Really
> cool!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tmrowco/northapp-contrib/pull/399#issuecomment-635036958>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMUIKENBOFDEZEBOJVXCALRTW447ANCNFSM4NDBD6JA>
> .
>


-- 

Olivier Corradi

Founder, CEO

<https://www.linkedin.com/in/oliviercorradi> <https://twitter.com/corradio>

tmrow.com
